### PR TITLE
Write New `.level` Files to Subdirectories

### PR DESCRIPTION
**Pull request by Hamms**
Follow-up to https://github.com/code-dot-org/code-dot-org/pull/52978

Out of a desire to reduce the number of files we are attempting to store in a single directory (currently over 30,000!), start writing new level files into a subdirectory based on the Game the Level is associated with, if possible. We also take this opportunity to move level files out of the `config/scripts` directory and into a new `config/levels` directory, and in preparation for eventually also reorganizing our DSL-defined level files we put all our custom level files into a `custom` directory.

Given our current set of levels, we'd end up with a directory structure that looks something like this:

```
dashboard/
  config/
    levels/
      custom/
        aichat/ (2 files)
        ailab/ (105 files)
        applab/ (6651 files)
        bounce/ (136 files)
        calc/ (57 files)
        craft/ (359 files)
        curriculum_reference/ (1340 files)
        dance/ (156 files)
        eval/ (105 files)
        external_link/ (64 files)
        fish/ (79 files)
        flappy/ (123 files)
        free_response/ (2615 files)
        frequency_analysis/ (24 files)
        gamelab/ (3637 files)
        javalab/ (3637 files)
        maze/ (3372 files)
        music/ (16 files)
        netsim/ (91 files)
        odometer/ (9 files)
        pixelation/ (156 files)
        poetry/ (233 files)
        public_key_cryptography/ (8 files)
        spritelab/ (2412 files)
        standalone_video/ (1623 files)
        studio/ (1106 files)
        text_compression/ (10 files)
        turtle/ (2147 files)
        unplug/ (126 files)
        vigenere/ (8 files)
        weblab/ (1205 files)
```

Also define a new helper method `Services::LevelFiles.reorganize_level_file_into_subdirectory(level)` which can move an existing level file into this new directory structure.

Couple details that are worth calling out:

I updated the helper method `Policies::LevelFiles.level_file_path` to take as its argument a Level model object rather than a string representing the name of a level. This change is representative of an underlying functional change to the way we think about level file paths: it is no longer possible to infer the desired default path to the level file from the name of the level alone. We now also require information about another model instance associated with the level object, and so need to consult the current state of the level in the database rather than being able to operate off of an identifier alone.

I updated `Services::LevelFiles.delete_custom_level_file(level)` to use the existing `level_file_path` helper rather than manually recreating a subset of its logic. I'm not sure why it wasn't doing that already; could be an optimization we just missed, but it's possible there's a reason I'm not seeing why we actually wanted these to be separate.

## Testing story

Updated and expanded existing unit tests to check for the new subdirectories and accommodate new edge cases.

Also tested manually by enabling levelbuilder mode on my local environment; verified that new level files are created in the desired location, that existing levels continue to update files in the old location for as long as they exist, and that existing levels will update files in the new location once moved.

## Follow-up work

1. Verify that newly-created levels are correctly serialized and deserialized
2. Verify that existing levels whose level files have been reorganized can continue to be updated
3. Uncomment the line in `Services::LevelFiles.write_custom_level_file` which will begin reorganizing levels that are under active development
4. Invoke `reorganize_level_file_into_subdirectory` on all remaining existing levels, probably in reasonably-sized batches.

### CODE REVIEW

**Review by davidsbailey — COMMENTED** [review]
The code changes look great! I have two qualms about the new location of the files:
1. it would be nice to keep custom level files separate from dsl-defined level files, for ease of use while code searching
2. I don't see any reason to keep `scripts` in the path

based on that, I would propose `dashboard/config/levels/custom` in this PR and `dashboard/config/levels/dsl_defined` whenever we split up the DSL-defined levels.

Thoughts?

File: dashboard/lib/policies/level_files.rb [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
-    # a path to which such a file should be written if it doesn't.
-    def self.level_file_path(level_name)
-      level_paths = Dir.glob(Rails.root.join("config/scripts/**/#{level_name}.level"))
-      raise("Multiple .level files for '#{level_name}' found: #{level_paths}") if level_paths.many?
-      level_paths.first || Rails.root.join("config/scripts/levels/#{level_name}.level")
+    # We organize new level files into a `levels` directory to keep them
+    # separate from scripts and further organize them by the associated Game if
+    # it has one, to avoid packing too many files into a single directory.
+    def self.default_level_file_path(level)
+      return Rails.root.join(*['config/scripts/levels', level.game&.app, "#{level.name}.level"].compact)
```
Comment (davidsbailey) [inline dashboard/lib/policies/level_files.rb]:
```suggestion
      return Rails.root.join(*['config/levels/custom', level.game&.app, "#{level.name}.level"].compact)
```

**Comment by Hamms** [comment]
I dig it! Moving all of these out of the scripts directory seems like something we probably should have done a long time ago, and planning ahead for DSL-defined levels definitely makes sense.

**Review by davidsbailey — APPROVED** [review]
LGTM!

**Review by bethanyaconnor — APPROVED** [review]
Love that we're getting rid of the /scripts path! Also thank you for setting up the ability to invoke `reorganize_level_file_into_subdirectory` in this PR but waiting until we can test it before automatically calling it. A bug that caused lost work for levelbuilders was definitely on my list of concerns :)

Change LGTM! Thanks again for taking this on

**Pull request merged**